### ISSUE TITLE
Subscriptions fire when mutable is set to same value

### ISSUE THREAD

**Issue by charlesroddie** [ReedCopsey/Gjallarhorn#52]
When a mutable is set to the same value as its current value, the subscriptions on it still fire, and mutables defined as maps of it are also set again, which causes subscriptions on them to fire.

This is not on the latest version of Gjallarhorn and I will update this post to confirm whether the issue still exists when we are able to upgrade to the latest.

**Comment by ReedCopsey** [comment]
If you can, make a repro in latest. Mutable uses `EqualityComparer<'a>.Default` to check for equality...

**Comment by charlesroddie** [comment]
This issue exists using the latest nuget package 1.1.0 and is related to Signal.map.

```fsharp
open Gjallarhorn

let x = Mutable.create (0.)
let y = x |> Signal.map (fun (f:float) -> int f)
do  Signal.Subscription.create (fun (i:int) -> System.Diagnostics.Debug.WriteLine i) y |> ignore
    for f in [0. .. 0.1 .. 1.] do
        x.Value <- f
```
Expected output: `1` (`01` would be sensible but subscriptions don't act on the current value, only changes)
Actual output: `0000000001`

Without the Signal.map we get `1`:
```fsharp
let z = Mutable.create 0
do  Signal.Subscription.create (fun (i:int) -> System.Diagnostics.Debug.Write i) z |> ignore
    for f in [0. .. 0.1 .. 1.] do
        z.Value <- int f
```

**Comment by ReedCopsey** [comment]
Fixed in 1.2.1 - https://www.nuget.org/packages/Gjallarhorn/1.2.1

See test at https://github.com/ReedCopsey/Gjallarhorn/blob/master/tests/Gjallarhorn.Tests/Signal.fs#L435-L443 to prevent this in the future.

**Comment by charlesroddie** [comment]
Thanks for this. There is some oddities remaining though. The following code is OK after your 1.2.1 fix:

```fsharp
let mutable calcs = 0
let x = Mutable.create 0
let y = x |> Signal.map ignore // y never changes,  so subscriptions on y should not fire
y |> Signal.Subscription.create (fun i -> calcs <- calcs + 1) |> ignore
do x.Value <- 1
calcs  // this gives 0 as expected
```
However replacing `y |> Signal.Subscription.create` with `y.Subscribe` gives 1. So there is some announcement of changes by Observable signals that don't change. I mention this as it may be connected to the following problem, which only uses Gjallarhorn methods:
```fsharp
let mutable calcs = 0
let x = Mutable.create 0
let y = x |> Signal.map ignore // y is always ()
let z = y |> Signal.map (fun x ->
    calcs <- calcs + 1) // so there is never a need to update z after initialization
z |> Signal.Subscription.create ignore |> ignore // this is to remove laziness
do x.Value <- 1
calcs // this should be 1
```
Actual output: 2
When x changes, y is updated to (). Even though y has not changed, it is apparently marked as "dirty", presumably as a map of something dirty. Therefore z is recalculated.

Thanks for your help and I hope you don't mind my variations on the same issue!

**Comment by ReedCopsey** [comment]
Fixed in 1.2.2
